Mobile: Native log in

olivaresf opened the pull request:
Adds JSON API support for the magic link authentication flow, enabling native mobile clients to authenticate. Instead of setting a cookie as an added security measure, we're sending back a `pending_authentication_token` that needs to be returned to the server when validating a code.

**POST /session.json { email_address: "..." }**
- Returns { pending_authentication_token: "..." } (201)
- Token expires in 10 minutes.
- If a magic link cannot be sent, the web redirects to `session_magic_link_path`, so there's no leak regarding email presence. For that reason, this endpoint always returns a `pending_authentication_token` even if no email was sent.
- If an invalid email is sent, we return `:unprocessable_entity` just like the web.

**POST /session/magic_link.json { code: "...", pending_authentication_token: "..." }**
- Returns { session_token: "..." } (200)
- Returns 401 if token missing, expired, invalid code, or email mismatch

olivaresf committed:
Allow JSON requests to send a magic link

olivaresf committed:
Allow JSON requests submitting a magic link code

olivaresf committed:
Dev: Set magic link as header when JSON request in development

olivaresf committed:
Return email and users too

olivaresf commented:
Thanks for your feedback! This is a WIP and isn't ready yet 😄

monorkin committed:
Cleanup & simplify sign in

monorkin commented:
@olivaresf seems like you don't have your email address added to Github so it doesn't show proper attribution on your commits.
(After rebasing, they look like I made them, but that should fix itself after you add your email address)

monorkin committed:
Remove email address and users from magic link response

The identity endpoint can be used to fetch that information

monorkin committed:
Add tests for sign in via API

monorkin commented on app/controllers/sessions/magic_links_controller.rb:
```diff
@@ -22,11 +21,41 @@ def code
+
+        format.json do
+          new_access_token = magic_link.identity.access_tokens.create!(permission: :write)
+          render json: { access_token: new_access_token.token }
```
Jay added an identity endpoint that returns the email, account, user and other info but using a jbuilder template so we get to levarage caching and control the response payload.

So I removed those fields from here. I know it's now two requests to get everything, hope that doesn't cause too many problems.

monorkin commented on app/controllers/sessions/magic_links_controller.rb:
```diff
@@ -22,11 +21,41 @@ def code
+        end
+
+        format.json do
+          new_access_token = magic_link.identity.access_tokens.create!(permission: :write)
```
Creating a token might not be what we want in this case because they'll show up in the access tokens section.
Maybe we could use the session cookie to authenticate the mobile apps? IDK, how we do auth on other apps, will check that. But the cookie would be the easiest solution.

monorkin committed:
Return the session cookie

We had a call about this. In short, we could reuse access tokens but then the user would see access tokens for every mobile device they have without any indication as to what is going on. So, since this really is just logging in instead of an integration which seems to be the primary purpose of access tokens, we can just use our regular session cookie for authentication.

monorkin committed:
Add back missing development magic link prompt

olivaresf committed:
Simplify session create logic for both html and json

olivaresf committed:
Simplify code a bit

olivaresf committed:
Pass a server token when creating a magic link via API

olivaresf committed:
Add unit tests for the new endpoints

olivaresf committed:
Change test expectation on single tenant mode account creation

olivaresf committed:
Update API test for cross code

olivaresf committed:
Update to always return a pending auth token for JSON responses.

olivaresf committed:
Cleanup session creation

olivaresf committed:
Rename test to clarify what they're about

olivaresf committed:
Move magic link api tests to their own files

olivaresf committed:
Restore sessions_controller test on creating invalid email address

olivaresf committed:
Fix due to unit test when creating with invalid emails

monorkin committed:
Simplify auth logic

monorkin committed:
Split tests by controller or responsibility

monorkin committed:
Clean up interfaces

I talked to the mobile team, and to keep things simple we agreed to send
the token via a cookie.

monorkin committed:
Tidy up session_token

monorkin committed:
Replace FakeMagicLink with a temporary object

kevinmcconnell reviewed (commented):
@monorkin I added a few thoughts re your recent changes. This all looks really good though!

olivaresf commented on app/controllers/concerns/authentication/via_magic_link.rb:
```diff
@@ -0,0 +1,67 @@
+    def serve_development_magic_link(magic_link)
+      if Rails.env.development? && magic_link.present?
+        flash[:magic_link_code] = magic_link.code
+        response.set_header("X-Magic-Link-Code", magic_link.code)
```
This is a convenience for when native apps are being tested against development. For example, an iOS simulator pointing to fizzy.localhost will receive the code via headers and can display it in-app. 😁

monorkin committed:
Use same constant for fake magic links

monorkin commented on app/controllers/concerns/authentication/via_magic_link.rb:
```diff
@@ -0,0 +1,67 @@
+        expires_at: 15.minutes.from_now
+      )
+
+      redirect_to_session_magic_link fake_magic_link, **options
```
Thank you for suggesting it!

monorkin committed:
Replace handle_ naming

monorkin committed:
Document the new sign in method
